### DISCUSSION TITLE
unsupervised learning for a hetero graph

### DISCUSSION THREAD

**Discussion by hanj99** [pyg-team/pytorch_geometric#3696] [Q&A] [answered]
if a graph is heterogeneous and want to do unsupervised learning, what would be the best approach?

What i'm thinking is as follows.

1. make the nearest node pairs similar in the embedding space, and distant node pairs dissimilar.
But, this procedure needs random walking from the specific node type to the same node type in n-hop away 
to get distant node pairs. AFIK, there is only homogeneous random walk api in the torch cluster. 
Any good idea to implement this approach?

2. predict whether link between the two nodes exist.
On different node type A and B, make a prediction about the existence of a link between A and B using their embeddings.

I'm concerned about which approach is better.
Thank you.

**Comment by rusty1s** [comment, marked as answer]
I think approach 2 is more straightforward to apply. In approach 1, it is unclear what a similar embedding actually means for a pair of nodes of different type. We also have PR https://github.com/pyg-team/pytorch_geometric/pull/3189 which introduces an unsupervised training method for heterogeneous graphs, but I haven't had a closer look yet. You can also try to convert "Deep Graph Infomax" to heterogeneous graphs.

**Reply by hanj99** [reply]
Thank you rusty. 
I thought if I implement using approach 1, pair of nodes would be the same type. But after thinking about it again, that might not be appropriate to compare their embeddings because there is always 2-hop distance (which is rather far) between the same types. 
For example, if there are 2 types (User, Article) and graph is User-(clicks)-Article-(clicked-by)-User, then comparison between Users might not be good to calculate their similarity.

So like you suggested, approach 2 seems more appropriate.
Thank you. And I'll look into what is "Deep Graph Infomax"
